File: controllers/portfolio.go
package controllers

import (
	"fmt"
	"github.com/gofiber/fiber"
	"github.com/lord-jerry/trading-history/database"
	"github.com/lord-jerry/trading-history/models"
	"github.com/lord-jerry/trading-history/utils"
)

func CreatePortfolio(c *fiber.Ctx) {
	db := database.Init()
	var findUser models.User
	decodedToken := utils.DecodeToken(c)

	db.Find(&findUser, int(decodedToken.UserID))
	if findUser.Email == "" {
		c.Status(404).JSON(fiber.Map{
			"message":    fmt.Sprintf("User with Id %s, does not exits", string(int(decodedToken.UserID))),
			"statusCode": 404,
		})
		return
	}

	portfolio := models.Portfolio{
		UserID: findUser.ID,
		Type:   c.FormValue("type"),
		Name:   c.FormValue("name"),
	}

	db.Create(&portfolio)
	c.Status(201).JSON(fiber.Map{
		"message":    "Portfolio created Successfully",
		"data":       portfolio,
		"statusCode": 201,
	})
}

func GetAllPortfolios(c *fiber.Ctx) {
	db := database.Init()
	var portfolio []models.Portfolio
	decodedToken := utils.DecodeToken(c)

	db.Where("user_id = ?", uint(decodedToken.UserID)).Find(&portfolio)
	c.Status(200).JSON(fiber.Map{
		"message":    "Portfolio Fetched Successfully",
		"data":       portfolio,
		"statusCode": 200,
	})
}

func EditPortfolio(c *fiber.Ctx) {
	db := database.Init()
	var portfolio models.Portfolio
	portfolioID := c.Params("id")
	decodedToken := utils.DecodeToken(c)

	db.Find(&portfolio, portfolioID)
	if portfolio.UserID != uint(decodedToken.UserID) {
		c.Status(404).JSON(fiber.Map{
			"message":    fmt.Sprintf("Portfolio with Id %s, does not exits", portfolioID),
			"statusCode": 404,
		})
		return
	}
	portfolio.Name = c.FormValue("name")
	portfolio.Type = c.FormValue("type")
	db.Save(&portfolio)

	c.Status(200).JSON(fiber.Map{
		"message":    "Portfolio updated Successfully",
		"data":       portfolio,
		"statusCode": 200,
	})
}

func DeletePortfolio(c *fiber.Ctx) {
	db := database.Init()
	var portfolio models.Portfolio
	portfolioID := c.Params("id")
	decodedToken := utils.DecodeToken(c)

	db.Find(&portfolio, portfolioID)
	if portfolio.UserID != uint(decodedToken.UserID) {
		c.Status(404).JSON(fiber.Map{
			"message":    fmt.Sprintf("Portfolio with Id %s, does not exits", portfolioID),
			"statusCode": 404,
		})
		return
	}

	db.Delete(&portfolio)
	c.Status(204)
}

File: server/src/controllers/portfolio.ts
import { Request, Response, NextFunction } from 'express';
import { UserInterface } from '../interfaces/User';
import PortfolioService from '../services/PortfolioService';

export default class Portfolio {
    static async create(req: Request, res: Response, next: NextFunction) {
        try {
            const { body } = req;
            const { id: userId } : UserInterface = Object(req.headers?.decoded_token);
            const result = await PortfolioService.createPortfolio(userId, body);

            return res.status(201).json({
                status: 201,
                message: 'Portfolio created successfully',
                data: {
                    ...result,
                },
            });
        } catch (err) {
            return next(err);
        }
    }

    static async getAll(req: Request, res: Response, next: NextFunction) {
        try {
            const { id: userId } : UserInterface = Object(req.headers?.decoded_token);
            const result = await PortfolioService.getAllPortfolios(userId);

            return res.status(200).json({
                status: 200,
                message: 'All Portfolios',
                data: result,
            });
        } catch (err) {
            return next(err);
        }
    }

    static async update(req: Request, res: Response, next: NextFunction) {
        try {
            const { id: userId } : UserInterface = Object(req.headers?.decoded_token);
            const portfolioId = Number(req.params?.portfolioId);
            const { body } = req;

            const result = await PortfolioService.updatePortfolio(userId, portfolioId, body);

            return res.status(200).json({
                status: 200,
                message: 'Portfolio updated',
                data: result,
            });
        } catch (err) {
            return next(err);
        }
    }

    static async delete(req: Request, res: Response, next: NextFunction) {
        try {
            const { id: userId } : UserInterface = Object(req.headers?.decoded_token);
            const portfolioId = Number(req.params?.portfolioId);

            await PortfolioService.deletePortfolio(userId, portfolioId);

            return res.status(204).json({
                status: 204,
                message: 'Portfolio deleted',
            });
        } catch (err) {
            return next(err);
        }
    }
}
